fml/endianness.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_ENDIANNESS_H_
+#define FLUTTER_FML_ENDIANNESS_H_
+
+#include <cstdint>
+#include <type_traits>
+#if defined(_MSC_VER)
+#include "intrin.h"
+#endif
+
+#include "flutter/fml/build_config.h"
+
+// Compiler intrinsics for flipping endianness.
+#define BYTESWAP16(n) __builtin_bswap16(n)
+#define BYTESWAP32(n) __builtin_bswap32(n)
+#define BYTESWAP64(n) __builtin_bswap64(n)
+
+#if defined(_MSC_VER)
+#define BYTESWAP16(n) _byteswap_ushort(n)
+#define BYTESWAP32(n) _byteswap_ulong(n)
+#define BYTESWAP64(n) _byteswap_uint64(n)
+#endif
+
+namespace fml {
+
+/// @brief  Flips the endianness of the given value.
+template <typename T>
+constexpr T ByteSwap(T n) {
```

Comment:
When using templating in C++, please be mindful of the error messages being generated. For instance, if I say `ByteSwap("Hello")`, the compiler will indicate an error in the header and not the location of incorrect call. Admittedly, there is a note that indicates what led to the error but the user needs to dig deeper to figure out whats going on and the first spot the compiler points you to as being the error is not actually the error.
<img width="518" alt="Screen Shot 2022-01-27 at 2 17 17 PM" src="https://user-images.githubusercontent.com/44085/151452883-3af4d3e4-1d48-4740-9f72-3e40f41f54d7.png">

One way to avoid this in the scheme you have devised would be to have restrictions on specialization. So, something like `, class = std::enable_if_t<std::is_integral<T>::value>`.

This makes the error move to the right spot.
<img width="716" alt="Screen Shot 2022-01-27 at 2 21 15 PM" src="https://user-images.githubusercontent.com/44085/151453318-956eb781-64a5-4e2c-8cfd-56d24cc69e38.png">

The compiler note will indicate why the specialization failed. Here, the error is in the right spot and the note indicates what the user must to do make the specialization work.

BTW, the restriction may even avoid you needing the check for other zero sized typed and the whole static assert can be `false`.

In case you [want a link to the sandbox](https://godbolt.org/z/ne7Ybzqj5) where you can experiment with templating and error messages.

Another suggestion would be to just add explicit overloads for each integral type.  I suppose that would work for ByteSwap but get tedious for `<Foo>EndianFrom/ToArch`. So, your call on that suggestion.

Templating is really interesting and powerful but its easy to call it a day without paying sufficient attention to UX concerns around its use which gives the technique a worse rep that it perhaps deserves.

Reply:
Interesting! I haven't touched adding type restrictions like this, so this was illuminating. I added what you suggested here, but if you have a preference for adding overloads in this case, let me know and I'll do it (I don't have a strong opinion either way).